data/src/test/java/com/microsoft/azure/kusto/data/AadAuthenticationHelperTest.java

```diff
@@ -83,4 +96,37 @@ static KeyCert readPem(String path, String password)
         return keycert;
     }
 
+    @Test
+    @DisplayName("validate cached token. Refresh if needed. Call regularly if no refresh token")
+    void useCachedTokenAndRefreshWhenNeeded() throws InterruptedException, ExecutionException, ServiceUnavailableException, IOException, DataServiceException, URISyntaxException, CertificateException, OperatorCreationException, PKCSException {
+        String certFilePath = Paths.get("src","test","resources", "cert.cer").toString();
+        String privateKeyPath = Paths.get("src","test","resources","key.pem").toString();
+
+        X509Certificate x509Certificate = readPem(certFilePath, "basic").getCertificate();
+        PrivateKey privateKey = readPem(privateKeyPath, "basic").getKey();
+
+        ConnectionStringBuilder csb = ConnectionStringBuilder
+                .createWithAadApplicationCertificate("resource.uri", "client-id", x509Certificate, privateKey);
+
+        AadAuthenticationHelper aadAuthenticationHelperSpy = spy(new AadAuthenticationHelper(csb));
+
+        AuthenticationResult authenticationResult = new AuthenticationResult("testType", "firstToken","refreshToken",0,"id", mock(UserInfo.class),false);
+        AuthenticationResult authenticationResultFromRefresh = new AuthenticationResult("testType", "fromRefresh",null,90,"id", mock(UserInfo.class),false);
+        AuthenticationResult authenticationResultNullRefreshTokenResult = new AuthenticationResult("testType", "nullRefreshResult",null,0,"id", mock(UserInfo.class),false);
+
+        doReturn(authenticationResultFromRefresh).when(aadAuthenticationHelperSpy).acquireAccessTokenByRefreshToken();
+        doReturn(authenticationResult).when(aadAuthenticationHelperSpy).acquireWithClientCertificate();
+
+        assertEquals("firstToken", aadAuthenticationHelperSpy.acquireAccessToken());
+        assertEquals("fromRefresh", aadAuthenticationHelperSpy.acquireAccessToken());
+        assertEquals("fromRefresh", aadAuthenticationHelperSpy.acquireAccessToken());
```

Comment:
Cannot this test be split to 2 tests?
if one of these assertEquals() fails, it won't continue to the last assertEquals().

Reply:
i dont think i should. the test checks the flow - for the first time it should return a token from the regular acquireToken function. than it will refresh it as the last token is expired and will . Now it validates that calling the acquireAccessToken still gets the second token .
And the last assert checks that now (after time passed) - if token expires it will reacquire a token using the regular acquireToken function again.

but should i add comments ?

Reply:
OK I understand now.. you might add some comments.